Add timeouts for commit hooks

DaRacci opened the pull request:
<!---
Thank you for contributing to GitUI! Please fill out the template below, and remove or add any
information as you feel necessary.
--->

This Pull Request closes #2546.

It changes the following:
- Adds timeouts to git hooks

I followed the checklist:
- [x] I added unittests
- [x] I ran `make check` without errors
- [x] I tested the overall application
- [x] I added an appropriate item to the changelog

DaRacci commented:
Opening early as a draft.

Initial changes are working and have tests added for them just need to figure out how we are going to implement a timeout value, if thats user configuration or a default impl.

DaRacci commented:
Was wanting to get some feedback on this before progressing further, is this the correct place to implement the timeout logic?

extrawurst commented:
This would impact other users like Gitbutler. @Byron what do you think about this change?

cruessler commented on asyncgit/src/sync/hooks.rs:
```diff
@@ -174,4 +280,53 @@ mod tests {
+
+		let res = hooks_pre_commit(
+			&root.to_str().unwrap().into(),
+			Duration::ZERO,
```
Should this be a non-zero value? At least, I was expecting a non-zero value based on the test’s name.

DaRacci replied:
you're right this should be a non zero, will get that fixed up

cruessler reviewed (commented):
From an implementation standpoint, this looks good. The only thing I would add is tests for non-zero timeouts (maybe that’s something you already planned or maybe I missed something).

DaRacci commented:
Also worth mentioning as its done in this PR, I've changed the test shell shebang to use `/usr/bin/env sh` as this is a more widely accepted and allows distributions like NixOS to run the tests.

DaRacci commented:
I've also noticed that while running tests that since i use the non POSIX Shell (nushell) that i need to prefix the command with SHELL=bash, i was wondering if it may be worth testing if the $SHELL is a known non compliant shell and trying to look for a default POSIX shell in the path?

Byron commented on asyncgit/src/sync/hooks.rs:
```diff
@@ -26,50 +30,148 @@ impl From<git2_hooks::HookResult> for HookResult {
+}
+
 /// see `git2_hooks::hooks_commit_msg`
 pub fn hooks_commit_msg(
```
Even though it's certainly fine to make this a breaking change, I think making new variants of these functions with `_with_timeout()` suffix could be considered.

DaRacci replied:
I've implemented variants in both git2-hooks and asyncgit.

Byron commented on asyncgit/src/sync/hooks.rs:
```diff
@@ -26,50 +30,148 @@ impl From<git2_hooks::HookResult> for HookResult {
+			None,
+		)),
+		Err(RecvTimeoutError::Disconnected) => {
+			unreachable!()
```
If `f()` panics, `tx` would drop which I *think* may actually trigger this.

DaRacci replied:
I've tested this and if a panic occurs in the thread it propagates down and the unreachable isn't reached.

Byron commented on asyncgit/src/sync/hooks.rs:
```diff
@@ -26,50 +30,148 @@ impl From<git2_hooks::HookResult> for HookResult {
+	match rx.recv_timeout(timeout) {
+		Ok(result) => result,
+		Err(RecvTimeoutError::Timeout) => Ok((
+			HookResult::NotOk("hook timed out".to_string()),
```
It would be nice if timeout-errors would be discoverable. Then it would need to be strongly typed.

DaRacci replied:
This is now strongly typed in `git2_hooks::HookResult` and `asyncgit::sync::hooks::HookResult`

DaRacci commented:
Now that the implementation is solid how do i go about with setting the timeout, how do we expose configurable value for the user and should there be a default timeout?

Byron commented on asyncgit/src/sync/hooks.rs:
```diff
@@ -26,50 +30,148 @@ impl From<git2_hooks::HookResult> for HookResult {
+
+	match rx.recv_timeout(timeout) {
+		Ok(result) => result,
+		Err(RecvTimeoutError::Timeout) => Ok((
```
I don't think this is a viable way of handling hook timeouts. After all, the thread that calls `f()` will keep running, and so will the hook program.
Depending on what it does, there can be strange interactions if the calling application keeps running hooks which pile up that way.

Instead, I think the timeout functionality would have to go into `git2-hooks`, and the one controlling the `std::process::Command` would have to use polling to eventually [`kill`](https://doc.rust-lang.org/std/process/struct.Child.html#method.kill) the hook process. This is similar to what a user would do in the terminal if something takes too long.

Byron reviewed (commented):
Thanks for letting me know, @extrawurst.

The changes seem to not apply to `git2-hooks`, so wouldn't affect GitButler in its current form. However, that would change if timeouts would move into `git2-hooks` as I seem to be proposing :).

DaRacci commented:
@Byron I've updated the implementation and it is now inside git2-hooks, I've added new functions suffixed with_timeout for each hook but the run_hooks function has a breaking change having the timeout parameter added to it.

DaRacci commented:
The last remaining thing to do is move the tests i had written over to the git2_hooks crate, even so the current tests in asyncgit should be able to cover most cases i think.

DaRacci commented:
I was also thinking about the initial issue where the problem was an infinite wait because the hook was waiting for input, maybe a solution could be when a hook is running a new window is shown with the stdio from the hook and also allowing stdin, not sure if that would work how im thinking though.

Byron commented on git2-hooks/src/hookspath.rs:
```diff
@@ -130,7 +136,28 @@ impl HookPaths {
 			)
-			.output()?;
+			.stdout(Stdio::piped())
+			.stderr(Stdio::piped())
```
This changes the semantics we had before, where `stdin` was null (see the docs of [`output`](https://doc.rust-lang.org/std/process/struct.Command.html#method.output)).
Doing so should also make it impossible for a hook to hang waiting on `stdin`. Hooks which still hang are probably buggy in another way.

Byron commented on git2-hooks/src/hookspath.rs:
```diff
@@ -130,7 +136,28 @@ impl HookPaths {
+				}
+
+				std::thread::yield_now();
+				std::thread::sleep(Duration::from_millis(10));
```
Doesn't `sleep()` imply a yield, among other things?

Byron replied:
One could think about a quadratic backoff (or a custom sequence) here where the first few milliseconds are checked more often, and as it takes longer, one checks less often.
This might also be overthinking it as 10ms seems fine.

Depending on how this is done, I think one *should* mention in the docs that the `timeout` may prolong the actual runtimes, right now by up to 10ms.

Byron commented on git2-hooks/src/lib.rs:
```diff
@@ -66,6 +67,10 @@ pub enum HookResult {
 		/// path of the hook that was run
 		hook: PathBuf,
 	},
+	TimedOut {
```
Docs are missing.
Probably the crate could use a `deny(missing_docs)` as well.

DaRacci replied:
I looked at adding this to the crate but there are a number of errors that are outside the scope of this PR.

Byron commented on git2-hooks/src/hookspath.rs:
```diff
@@ -106,7 +108,11 @@ impl HookPaths {
 	/// this function calls hook scripts based on conventions documented here
 	/// see <https://git-scm.com/docs/githooks>
-	pub fn run_hook(&self, args: &[&str]) -> Result<HookResult> {
+	pub fn run_hook(
```
I still think this doesn't have to be a breaking change unless one really wants to tell folks to deal with the timeout. That timeouts are implemented now probably means most people didn't have issues with this (or solved them by fixing the hooks), so I'd argue that `run_hook_with_timeout()` should be favored.

Byron commented on git2-hooks/src/hookspath.rs:
```diff
@@ -106,7 +108,11 @@ impl HookPaths {
 
 	/// this function calls hook scripts based on conventions documented here
 	/// see <https://git-scm.com/docs/githooks>
```
Needs a documentation update.

Byron reviewed (commented):
Thanks a lot for the quick turnaround!

The new implementation definitely addresses the previous issue, and now the timeout has an actual effect on the hook.

When looking at `git2-hooks` exclusively, I am not sure I saw a test for this though.

DaRacci commented:
I'll get some tests put into the git2-hooks crate and resolve these issue when I get some times on the weekend to do so, please let me know if there is anything else that should be looked into for when i get to it.

Byron commented:
> @Byron I've updated the implementation and it is now inside git2-hooks, I've added new functions with suffixed `_with_timeout` for each hook but the `run_hooks` function has a breaking change having the timeout parameter added to it.

The idea would be to leave the `run_hooks()` function alone, and have it call the `_with_timeout()` function internally with a zero-timeout. That way one trivially avoids the breaking change while providing the feature to those who are interested in it.

> I was also thinking about the initial issue where the problem was an infinite wait because the hook was waiting for input, maybe a solution could be when a hook is running a new window is shown with the stdio from the hook and also allowing stdin, not sure if that would work how im thinking though.

I don't think this can be a solution, as hooking up `stdin` must be controlled by the caller. For now, I'd think that hooks inherit stdin from the calling process, and that should work fine even if these have no stdin, as hooks can detect this and even if not they would never be able to hang - reading from a closed file will immediately fail.

Nonetheless I agree that timeouts for hooks may be solving for 'buggy' hooks, and it might be better to fix those (or understand precisely why they maybe hang legitimately so that can be fixed on the side of `git2-hooks` instead).

DaRacci commented:
> The idea would be to leave the `run_hooks()` function alone, and have it call the `_with_timeout()` function internally with a zero-timeout. That way one trivially avoids the breaking change while providing the feature to those who are interested in it.

I've made this change so run_hooks is left alone functionally and there is now the with_timeout version.

> I don't think this can be a solution, as hooking up `stdin` must be controlled by the caller. For now, I'd think that hooks inherit stdin from the calling process, and that should work fine even if these have no stdin, as hooks can detect this and even if not they would never be able to hang - reading from a closed file will immediately fail.
> 
> Nonetheless I agree that timeouts for hooks may be solving for 'buggy' hooks, and it might be better to fix those (or understand precisely why they maybe hang legitimately so that can be fixed on the side of `git2-hooks` instead).

I think for now its best to track this in a new issue / PR, this one is pretty complete and ready to go as it is.

extrawurst commented:
@DaRacci please update with master

naseschwarz reviewed (changes requested):
Hi @DaRacci. Thank you for your work on this, I appreciate it. This all seems well though-out and usable.

I do have some points regarding some details of your implementation I'd like to discuss before I give the go-ahead.

DaRacci commented:
Apologies, I'm in the middle of building & swapping to a new computer I'll get back to this when I'm able to.

DaRacci commented on git2-hooks/src/lib.rs:
```diff
@@ -247,7 +346,7 @@ mod tests {
 		assert_eq!(res, HookResult::NoHookFound);
 
-		let hook = b"#!/bin/sh
+		let hook = b"#!/usr/bin/env sh
```
You are right NixOS does provide a link to /bin/sh by default but I've gotten into the habit of always rewriting it to `/usr/bin/env` as it just works for every interpreter, I can change this back if it causes issues.

DaRacci commented on src/popups/options.rs:
```diff
@@ -99,6 +102,19 @@ impl OptionsPopup {
+			"Timeout",
+			&self.options.borrow().hook_timeout().map_or_else(
+				|| "None".to_string(),
+				|d| format!("{d:?}"),
```
I was thinking the same thing originally but writing some redundant code or adding a new library to handle it seems like overkill for a single use.

DaRacci commented on git2-hooks/src/lib.rs:
```diff
@@ -144,34 +161,75 @@ pub fn hooks_commit_msg(
+/// Git hook: `commit_msg`
+///
+/// See [`hooks_commit_msg`] for more details.
+pub fn hooks_commit_msg_with_timeout(
```
I agree unifying making these call the timeout variants does make sense to ensure any changes aren't missed, However you mention a unified interface for git2-hooks, are you implying i should remove the non timeout variants to only leave the timeout versions with an Option Duration in git2-hooks?

DaRacci commented on git2-hooks/src/hookspath.rs:
```diff
@@ -135,48 +138,186 @@ impl HookPaths {
+///         Total Sleep: 140 milliseconds
+///
+/// Attempt 8:
+//     Sleep Time=(8^2)×1=64, capped by `MAX_SLEEP_MILLIS` of 50
+//          Actual Sleep: 50 milliseconds
+//          Total Sleep: 190 milliseconds
```
I think leaving it were it is makes sense as IDE's will pick up on it when working with the function.

naseschwarz commented on asyncgit/src/sync/hooks.rs:
```diff
@@ -224,4 +296,106 @@ mod tests {
+		let file = temp_dir.path().join("test");
+		let hook = format!(
+			"#!/usr/bin/env sh
+sleep 1
```
Ideally, you'll never hit those 60 s, right, as the timeout would cancel the test before that.

DaRacci replied:
You're correct, I had thought this was one of the tests without a timeout.

DaRacci commented on git2-hooks/src/hookspath.rs:
```diff
@@ -290,4 +432,53 @@ mod test {
+	/// does not cause the total execution time does not grealy increase the total execution time.
+	#[test]
+	fn test_timeout_with_quadratic_backoff_cost() {
+		let timeout = Duration::from_millis(100);
+		let start = std::time::Instant::now();
+		let result =
+			timeout_with_quadratic_backoff(timeout, || Ok(false));
+		let elapsed = start.elapsed();
+
+		assert_eq!(result.unwrap(), false);
+		assert!(elapsed < timeout + Duration::from_millis(10));
```
I've honestly got no idea how to make my own Instant mock or sans-i/o implementation, I've relaxed some of the test timings some more however.

DaRacci commented on git2-hooks/src/hookspath.rs:
```diff
@@ -135,48 +138,186 @@ impl HookPaths {
+			})? {
+				debug!("killing hook process");
+				child.kill()?;
+				return Ok(HookResult::TimedOut { hook });
```
I've implemented this with a process group so it actually kills and doesn't wait on under unix but I've got no idea where to start with killing a process tree under windows so it will still wait for children to finish under windows

DaRacci commented:
I've rebased and addressed most of the suggestions but there's some tests that seems to be failing at random that are unrelated to my changes as they still randomly fail or pass even on the master branch for me.
